mhs-reference-implementation/mhs/parser/ebxml_message_parser.py

```diff
@@ -1,38 +1,166 @@
+REF_TO_MESSAGE_ID = "ref_to_message_id"
+DUPLICATE_ELIMINATION = "duplicate_elimination"
+ACK_REQUESTED = "ack_requested"
+ACK_SOAP_ACTOR = "ack_soap_actor"
+SYNC_REPLY = "sync_reply"
+
+MESSAGE = "message"
+
+EBXML_NAMESPACE = "eb"
+SOAP_NAMESPACE = "SOAP"
+NAMESPACES = {SOAP_NAMESPACE: "http://schemas.xmlsoap.org/soap/envelope/",
+              EBXML_NAMESPACE: "http://www.oasis-open.org/committees/ebxml-msg/schema/msg-header-2_0.xsd"}
+
+CONTENT_TYPE_HEADER_NAME = "Content-Type"
+
+
+class EbXmlParsingError(Exception):
+    """Raised when an error was encountered during parsing of an ebXML message."""
+    pass
 
 
 class EbXmlMessageParser:
     """A component that extracts information from ebXML messages."""
 
-    def parse_message(self, message):
+    def parse_message(self, headers, message):
         """Parse the provided ebXML message and extract a dictionary of values from it.
 
+        :param headers A dictionary of headers received with the message.
         :param message: The message to be parsed.
         :return: A dictionary of values extracted from the message.
         """
-        # TODO: Implement schematron/XSLT parsing here.
-        pass
+        xml_tree = ET.fromstring(message)
+        extracted_values = {}
 
+        self._add_if_present(extracted_values, FROM_PARTY_ID,
+                             self._extract_ebxml_text_value(xml_tree, "PartyId", parent="From"))
+        self._add_if_present(extracted_values, TO_PARTY_ID,
+                             self._extract_ebxml_text_value(xml_tree, "PartyId", parent="To"))
+        self._add_if_present(extracted_values, CPA_ID, self._extract_ebxml_text_value(xml_tree, "CPAId"))
+        self._add_if_present(extracted_values, CONVERSATION_ID,
+                             self._extract_ebxml_text_value(xml_tree, "ConversationId"))
+        self._add_if_present(extracted_values, SERVICE, self._extract_ebxml_text_value(xml_tree, "Service"))
+        self._add_if_present(extracted_values, ACTION, self._extract_ebxml_text_value(xml_tree, "Action"))
+        self._add_if_present(extracted_values, MESSAGE_ID,
+                             self._extract_ebxml_text_value(xml_tree, "MessageId", parent="MessageData"))
+        self._add_if_present(extracted_values, TIMESTAMP,
+                             self._extract_ebxml_text_value(xml_tree, "Timestamp", parent="MessageData"))
+        self._add_if_present(extracted_values, REF_TO_MESSAGE_ID,
+                             self._extract_ebxml_text_value(xml_tree, "RefToMessageId", parent="MessageData"))
+        self._add_flag_if_present(extracted_values, DUPLICATE_ELIMINATION,
+                                  self._extract_ebxml_value(xml_tree, "DuplicateElimination"))
+        self._add_flag_if_present(extracted_values, SYNC_REPLY, self._extract_ebxml_value(xml_tree, "SyncReply"))
+        self._add_flag_if_present(extracted_values, ACK_REQUESTED, self._extract_ebxml_value(xml_tree, "AckRequested"))
```

Comment:
I am just wondering if some of this can be extracted into some kind of parsing config. With some details about where to extract the relevant bits of data from. Then perhaps just iterating through this config to generate this extracted_value dict. Then it should be a case of updating this config if anything else is needed in the future.

Reply:
As discussed offline, I've created a tech-debt ticket to do this, as I think it's probably a little beyond the scope of this task.

Reply:
I was wondering something similar, seems like there are quite a few similar operations happening here which makes me feel like this should be looped somehow
Also quite a few of the constants used here are the only occurrence of these constants, seeing a there are already some fixed strings here with only 1 use do we need to keep all the constants at the top?

Reply:
I didn't refresh the page till I'd posted this so didn't see the previous comment but that's reasonable

Reply:
Regarding the constants, the idea was that they should be available for import from other modules, so that you can use them to index the dictionary returned from `parse_message`.